pilot/pkg/model/context.go

```diff
@@ -248,6 +253,43 @@ func GetProxyConfigNamespace(proxy *Proxy) string {
 	return parts[1]
 }
 
+// SetNamespaceDependencies sets the namespaces of the reachable services.
+func (node *Proxy) SetNamespaceDependencies(dependency *meshconfig.MeshConfig_DefaultServiceDependency) {
+	if node == nil || dependency == nil {
+		return
+	}
+	node.NamespaceDependencies = make(map[string]bool)
+	if dependency.ImportMode == meshconfig.MeshConfig_DefaultServiceDependency_ALL_NAMESPACES {
+		return
+	}
+
+	if dependency.ImportMode == meshconfig.MeshConfig_DefaultServiceDependency_SAME_NAMESPACE {
+		node.NamespaceDependencies[node.ConfigNamespace] = true
+		for _, ns := range dependency.ImportNamespaces {
+			node.NamespaceDependencies[ns] = true
+		}
+	}
+}
+
+// CanImport returns whether the node can import config of namespace 'ns' with ConfigScope 'scope'.
+func (node *Proxy) CanImport(ns string, scope networking.ConfigScope) bool {
```

Comment:
I'm not sure this is the right pattern - it'll still be quite expensive to compute it every time. 

All configs for a node ( serviceEnties, dependencies, etc) should be in the Proxy struct and updated:
- on connect
- when a config changes - triggering a push if that node has to change

This is needed for the upcoming 'on-demand', and to finally move to a 'push to a node only if a config affecting the node has changed' ( which is also critical for scalability)

That's one of the reasons we're moving stuff into the node.

Reply:
Yes, although this function is very efficient just have several judgement, but during push it has to be called O(n). 

I agree this is a little more expensive than caching the dependencies.  Would optimize later.

Reply:
Let's have proper data structures and design first - we already went down 'optimize later' path for everything else, and we're still paying the price. 

I would also point that making something public doesn't mean it has to be forced as import everywhere - the 'private'/public is just a mean to prevent import. We first need to determine the dependencies - and exclude non-public. 

Public/private is nice - but not directly affecting scalability - so no time pressure, can be done in 1.2

Reply:
agree

Reply:
>All configs for a node ( serviceEnties, dependencies, etc) should be in the Proxy struct and updated:
> on connect
>when a config changes - triggering a push if that node has to change

This is very inefficient and is quite unnecessary as well. The config scoping was part of the whole Pr for service dependency and that design doc. Public is the default, same behavior as today. And it implies that the config has to be applicable at all client nodes (as its done today). We cant change that semantic unless we want to do a breaking change.